skiko/src/iosMain/kotlin/org/jetbrains/skiko/redrawer/MetalRedrawer.ios.kt

```diff
@@ -1,25 +1,142 @@
+
+    private val frameDispatcher = FrameDispatcher(AppDispatcher) {
+        if (layer.isShowing()) {
+            update(getTimeNanos())
+            draw()
+        }
+    }
+
+    fun makeContext() = DirectContext.makeMetal(device.objcPtr(), queue.objcPtr())
+
+    fun makeRenderTarget(width: Int, height: Int): BackendRenderTarget {
+        currentDrawable = metalLayer.nextDrawable()!!
+        return BackendRenderTarget.makeMetal(width, height, currentDrawable!!.texture.objcPtr())
+    }
+
     override fun dispose() {
-        TODO()
+        frameDispatcher.cancel()
+        isDisposed = true
     }
 
     override fun syncSize() {
-        TODO()
+        println("TODO: implement syncSize()")
     }
 
     override fun needRedraw() {
-        TODO()
+        check(!isDisposed) { "MetalRedrawer is disposed" }
+        frameDispatcher.scheduleFrame()
     }
 
     override fun redrawImmediately() {
-        TODO()
+        check(!isDisposed) { "MetalRedrawer is disposed" }
+        metalLayer.setNeedsDisplay()
+        update(getTimeNanos())
+        performDraw()
+    }
+
+    private fun performDraw() {
+        if (!isDisposed) {
+            layer.draw()
+        }
+    }
+
+    private fun update(nanoTime: Long) {
+        layer.update(nanoTime)
+    }
+
+    private fun draw() {
+        // TODO: maybe make flush async as in JVM version.
+        autoreleasepool {
+            performDraw()
+        }
+    }
+
+    fun finishFrame() {
+        autoreleasepool {
+            currentDrawable?.let {
+                val commandBuffer = queue.commandBuffer()!!
+                commandBuffer.label = "Present"
+                commandBuffer.presentDrawable(it)
+                commandBuffer.commit()
+                currentDrawable = null
+            }
+        }
+    }
+}
+
+class MetalLayer(
+    private val skiaLayer: SkiaLayer,
+    theDevice: MTLDeviceProtocol
+) : CAMetalLayer() {
+    init {
+        this.setNeedsDisplayOnBoundsChange(true)
+        this.removeAllAnimations()
+        // TODO: looks like a bug in K/N interop.
+        this.device = theDevice as objcnames.protocols.MTLDeviceProtocol?
+        this.pixelFormat = MTLPixelFormatBGRA8Unorm
+        this.contentsGravity = kCAGravityTopLeft
+        doubleArrayOf(0.0, 0.0, 0.0, 0.0).usePinned {
+            this.backgroundColor =
+                CGColorCreate(CGColorSpaceCreateDeviceRGB(), it.addressOf(0))
+        }
+        this.opaque = true
+        this.frame = skiaLayer.view.frame
+        skiaLayer.view.layer.addSublayer(this)
+    }
+
+    fun draw()  {
+        skiaLayer.update(getTimeNanos())
+        skiaLayer.draw()
+    }
+
+    fun dispose() {
+        this.removeFromSuperlayer()
+        // TODO: anything else to dispose the layer?
+    }
+
+    @Suppress("unused")
+    private fun performDraw() {
+        try {
+            draw()
+        } catch (e: Throwable) {
+            e.printStackTrace()
```

Comment:
Better not to eat exceptions here.

Fallback won't work when we implement it, and log will be polluted if the exception happened during animation.